### PR TITLE
specify restore height by YYYY-MM-DD format

### PR DESCRIPTION
**Pull request by kenshi84**
This PR uses `MAKE_CORE_RPC_VERSION` introduced in PR #1578. It also defines RPC version 1.6 and should be merged after PR #1568 which defines RPC version 1.5.

Example:

    ./monero-wallet-cli --generate-from-view-key watchonly_wallet
    Standard address: 47RH3d1JBXyhv3u3baLzGU6Z1G2F4UVMiDHZUYJEQqsVRViEpBV76t7eE11RsAFj8Z6v1wtchFWyHXEnyrG5fBSc6C2zPCH
    View key: 163eda7fade793eb40ce9fd39e3b20d6c5165476861aded7c4ded1a252d9e205
    Enter a password for your new wallet: 
    Confirm Password: 
    Generated new wallet: 47RH3d1JBXyhv3u3baLzGU6Z1G2F4UVMiDHZUYJEQqsVRViEpBV76t7eE11RsAFj8Z6v1wtchFWyHXEnyrG5fBSc6C2zPCH
    Restore from specific blockchain height (optional, default 0),
    or alternatively from specific date (YYYY-MM-DD): 2017-01-13
    Restore height is: 1222233
    Is this okay?  (Y/Yes/N/No): y
    Starting refresh...
    Refresh done, blocks received: 1104                             
    Balance: 0.000000000000, unlocked balance: 0.000000000000
    Background refresh thread started
    [wallet 47RH3d]:

### CODE REVIEW

File: src/rpc/core_rpc_server.cpp [hunk 1]
```diff
@@ -190,6 +190,27 @@ namespace cryptonote
...
   //------------------------------------------------------------------------------------------------------------------------------
+  bool core_rpc_server::on_get_blocks_by_height(const COMMAND_RPC_GET_BLOCKS_BY_HEIGHT::request& req, COMMAND_RPC_GET_BLOCKS_BY_HEIGHT::response& res)
+  {
+    CHECK_CORE_BUSY();
+    res.status = "Failed";
+    res.blocks.clear();
+    res.blocks.reserve(req.heights.size());
+    for (uint64_t height : req.heights)
+    {
+      block blk = m_core.get_blockchain_storage().get_db().get_block_from_height(height);
```
Comment (moneromooo-monero) [inline src/rpc/core_rpc_server.cpp]: This can throw, so you may be able to put the daemon in some weird state by asking for a non existent height. I think try try/catch/fail there would ensure this cannot happen.

File: src/simplewallet/simplewallet.cpp [hunk 2]
```diff
@@ -1090,6 +1074,71 @@ bool simple_wallet::init(const boost::program_options::variables_map& vm)
       bool r = new_wallet(vm, m_recovery_key, m_restore_deterministic_wallet, m_non_deterministic, old_language);
       CHECK_AND_ASSERT_MES(r, false, tr("account creation failed"));
     }
+    if (!m_restore_height && m_restoring)
+    {
+      uint32_t version;
+      if (!try_connect_to_daemon(false, &version))
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: This means you can't restore offline anymore. Maybe it should only do this if supplying a date rather than a block number.

File: src/simplewallet/simplewallet.cpp [hunk 3]
```diff
@@ -1090,6 +1074,71 @@ bool simple_wallet::init(const boost::program_options::variables_map& vm)
...
+            day   = boost::lexical_cast<uint16_t>(heightstr.substr(8,2));
+            m_restore_height = m_wallet->get_blockchain_height_by_date(year, month, day);
+            success_msg_writer() << tr("Restore height is: ") << m_restore_height;
+            std::string confirm = command_line::input_line(tr("Is this okay?  (Y/Yes/N/No): "));
+            if (std::cin.eof())
+              return false;
+            if(command_line::is_yes(confirm))
+              break;
+          }
+          catch (boost::bad_lexical_cast &)
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: Please make those const

File: src/simplewallet/simplewallet.cpp [hunk 4]
```diff
@@ -1090,6 +1074,71 @@ bool simple_wallet::init(const boost::program_options::variables_map& vm)
...
+            continue;
+          }
+          uint16_t year;
+          uint8_t month;  // 1, 2, ..., 12
+          uint8_t day;    // 1, 2, ..., 31
+          try
+          {
+            year  = boost::lexical_cast<uint16_t>(heightstr.substr(0,4));
+            month = boost::lexical_cast<uint16_t>(heightstr.substr(5,2));
+            day   = boost::lexical_cast<uint16_t>(heightstr.substr(8,2));
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: Type mismatches with above. You could also just use sscanf, a lot simpler,

File: src/simplewallet/simplewallet.cpp [hunk 5]
```diff
@@ -1090,6 +1074,71 @@ bool simple_wallet::init(const boost::program_options::variables_map& vm)
...
+        try
+        {
+          m_restore_height = boost::lexical_cast<uint64_t>(heightstr);
+          break;
+        }
+        catch (boost::bad_lexical_cast &)
+        {
+          if (version < 0x1006)
+          {
+            fail_msg_writer() << tr("bad m_restore_height parameter: ") << heightstr;
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: That's an incorrect error message, since the test is whether we're connected to a recent enough server.

Also, you want 10006, not 1006.

File: src/wallet/wallet2.cpp [hunk 6]
```diff
@@ -5297,6 +5297,95 @@ bool wallet2::parse_uri(const std::string &uri, std::string &address, std::strin
...
 }
 //----------------------------------------------------------------------------------------------------
+uint64_t wallet2::get_blockchain_height_by_date(uint16_t year, uint8_t month, uint8_t day)
+{
+  uint32_t version;
+  if (!check_connection(&version))
+  {
+    throw std::runtime_error("failed to connect to daemon: " + get_daemon_address());
+  }
+  if (version < 0x1006)
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: 10006 too. Maybe add a MAKE_RPC_VERSION(x,y) (((x)<<16)|(y)), and make CORE_RPC_VERSION use this. This will prevent bugs like this.

File: src/wallet/wallet2.cpp [hunk 7]
```diff
@@ -5297,6 +5297,95 @@ bool wallet2::parse_uri(const std::string &uri, std::string &address, std::strin
...
+  }
+  std::tm date = { 0, 0, 0, 0, 0, 0, 0, 0 };
+  date.tm_year = year - 1900;
+  date.tm_mon  = month - 1;
+  date.tm_mday = day;
+  if (date.tm_mon < 0 || 11 < date.tm_mon || date.tm_mday < 1 || 31 < date.tm_mday)
+  {
+    throw std::runtime_error("month or day out of range");
+  }
+  uint64_t timestamp_target = std::mktime(&date);
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: Check return value (makes the check above kinda redundant too)

File: src/wallet/wallet2.cpp [hunk 8]
```diff
@@ -5297,6 +5297,95 @@ bool wallet2::parse_uri(const std::string &uri, std::string &address, std::strin
...
+    uint64_t timestamp_mid = blk_mid.timestamp;
+    uint64_t timestamp_max = blk_max.timestamp;
+    if (!(timestamp_min <= timestamp_mid && timestamp_mid <= timestamp_max))
+    {
+      std::ostringstream oss;
+      oss << "returned timestamps are chronologically inconsistent: "
+          << timestamp_min << ", "
+          << timestamp_mid << ", "
+          << timestamp_max;
+      throw std::runtime_error(oss.str());
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: I think they actually can in practice, as they're set by miners, and have some window of leeway. It's fine though, since you don't really want the exact block, but something close enough, *before* the right block.

File: src/wallet/wallet2.cpp [hunk 9]
```diff
@@ -5297,6 +5297,95 @@ bool wallet2::parse_uri(const std::string &uri, std::string &address, std::strin
...
+    }
+    if (timestamp_target > timestamp_max)
+    {
+      throw std::runtime_error("specified date is in the future");
+    }
+    if (timestamp_target <= timestamp_min)
+    {
+      return height_min;
+    }
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: For the same reason, I would pick a "buffer" amount of blocks (not going below genesis block). At least a day or two, since your resolution is a day, and you don't want timezones to matter here, and people will enter dates in their timezone.

File: src/rpc/core_rpc_server.cpp (hunk 1 above)
Comment (kenshi84) [inline src/rpc/core_rpc_server.cpp]: ok

File: src/simplewallet/simplewallet.cpp (hunk 4 above)
Comment (kenshi84) [inline src/simplewallet/simplewallet.cpp]: It's intentional. `boost::lexical_cast` considers `uint8_t` as the character type and throws exception when trying to convert from multi-character strings. A cleaner way would be to put `static_cast<uint8_t>` in front of the cast - should I do so?

I prefer `lexical_cast` since I'm not confident in using `sscanf` properly. Also it's never used in the Monero source code, only used in some of the external libraries (LMDB, unbound and miniupnpc). I found [a comparison among many choices in C++](http://www.kumobius.com/2013/08/c-string-to-int/), and `lexical_cast` seems like a fairly good choice. It's also universally used in the Monero source code.

File: src/simplewallet/simplewallet.cpp (hunk 3 above)
Comment (kenshi84) [inline src/simplewallet/simplewallet.cpp]: ok

File: src/simplewallet/simplewallet.cpp (hunk 5 above)
Comment (kenshi84) [inline src/simplewallet/simplewallet.cpp]: I believe the error message is correct, since the user is either offline or connected to an older RPC (where the date format is unsupported) and is passing whatever string that doesn't parse as the block height. If we're not connected to a recent enough server and pass a string "2017-01-13", it should be considered as a bad parameter.

File: src/wallet/wallet2.cpp (hunk 6 above)
Comment (kenshi84) [inline src/wallet/wallet2.cpp]: ok

File: src/simplewallet/simplewallet.cpp (hunk 2 above)
Comment (kenshi84) [inline src/simplewallet/simplewallet.cpp]: ok

File: src/simplewallet/simplewallet.cpp (hunk 5 above)
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: An odd choice, but that's subjective I guess, so ok

File: src/simplewallet/simplewallet.cpp (hunk 4 above)
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: That's fine. For scanf, you'd use, eg:
if (sscanf(heightstr.c_str(), "%04u-%02u-%02u", &year, &month, &day) < 3) { error }
That seems a lot simple to me than a dozen lines, but your code seems alright otherwise, so it's fine. No need for static casts.

File: src/simplewallet/simplewallet.cpp (hunk 5 above)
Comment (kenshi84) [inline src/simplewallet/simplewallet.cpp]: Why do you think it's odd? I'm curious.

Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: Because it seems to not match the logic, as I've said in my previous comment. The parameter is correct, but you can't tell what it corresponds to. That's like saying "green" is not a correct color in the sentence "this house is green" just because you can't see the house.

File: src/wallet/wallet2.cpp (hunk 6 above)
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: This one ought to be a separate patch if you can. It might be useful in its own right.

File: src/simplewallet/simplewallet.cpp (hunk 5 above)
Comment (kenshi84) [inline src/simplewallet/simplewallet.cpp]: I think it does match the logic. If the connected daemon is old and doesn't support this feature, the wallet displays
> Restore from specific blockchain height (optional, default 0):

In this context, any string parameter like "2017-01-13" that doesn't parse as block height makes no sense (since the user is not told to be able to do so) and hence should be considered incorrect.

Only if the wallet is connected to the newer RPC, it displays
> Restore from specific blockchain height (optional, default 0),
or alternatively from specific date (YYYY-MM-DD):

and the user has the choice of using a block height or a date as the parameter.

File: src/wallet/wallet2.cpp (hunk 6 above)
Comment (kenshi84) [inline src/wallet/wallet2.cpp]: ok

File: src/wallet/wallet2.cpp (hunk 7 above)
Comment (kenshi84) [inline src/wallet/wallet2.cpp]: I'm not sure what you intend here. My intention was to reject inputs like "2016-77-88".

File: src/simplewallet/simplewallet.cpp (hunk 4 above)
Comment (kenshi84) [inline src/simplewallet/simplewallet.cpp]: I think it's better to add `static_cast` to indicate that the type mismatch is intentional.

File: src/wallet/wallet2.cpp (hunk 8 above)
Comment (kenshi84) [inline src/wallet/wallet2.cpp]: Can I assume that when such time reversal occurs, the corresponding block heights are close to each other and thus the algorithm returns the smallest height?

File: src/wallet/wallet2.cpp (hunk 9 above)
Comment (kenshi84) [inline src/wallet/wallet2.cpp]: ok

File: src/wallet/wallet2.cpp (hunk 6 above)
Comment (kenshi84) [inline src/wallet/wallet2.cpp]: Was it appropriate that I made PR #1578? Or should I make the patch in this PR as a separate commit?

File: src/wallet/wallet2.cpp (hunk 7 above)
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: mktime will presumably reject invalid dates. Your tests above reject widely invalid ones, but not all of them. Letting mktime perform the weeding allows you to reject, eg, 2012-04-31, while not having to add custom code to try and get them all.

Comment (kenshi84) [inline src/wallet/wallet2.cpp]: I'm not so familiar with these time-related functions, and the cppreference's [sample code](http://en.cppreference.com/w/cpp/chrono/c/mktime) shows:

    std::time_t t = std::time(NULL);
    std::tm tm = *std::localtime(&t);
    std::cout << "Today is           " << std::put_time(&tm, "%c %Z")
              << " and DST is " << (tm.tm_isdst ? "in effect" : "not in effect") << '\n';
    tm.tm_mon -= 100;  // tm_mon is now outside its normal range
    std::mktime(&tm);  // tm_dst is not set to -1; today's DST status is used
    std::cout << "100 months ago was " << std::put_time(&tm, "%c %Z")
              << " and DST was " << (tm.tm_isdst ? "in effect" : "not in effect") << '\n';

So, out-of-range values specified to any fields of `tm` will not cause `mktime` to produce error, instead it'll normalize the values. In other words, it won't reject 2012-04-31, but instead interprets it as 2012-05-01.

Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: Hmm. I'm not sure that's a good idea or not. I tend to think not... But OK.

Comment (kenshi84) [inline src/wallet/wallet2.cpp]: Why do you tend to think it (i.e. the current code) is not a good idea? Do you see any potential issues?

Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: I meant, I do not think the mktime behavior of "normalizing" is a good idea. I guess it helps "find the next date", but it also prevents validation. Your code's fine, in view of this behavior, which I did not realize.

**Comment by fluffypony** [comment]
Please rebase

**Comment by moneromooo-monero** [comment]
ok

**Comment by kenshi84** [comment]
rebased
